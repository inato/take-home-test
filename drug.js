export class Drug {
  constructor(name, expiresIn, benefit) {
    this.name = name;
    this.expiresIn = expiresIn;
    this.benefit = benefit;
  }
}
