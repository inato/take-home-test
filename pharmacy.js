export class Pharmacy {
  constructor(drugs = []) {
    this.drugs = drugs;
  }
  updateBenefitValue() {
    for (var i = 0; i < this.drugs.length; i++) {
      if (
        this.drugs[i].name != "Herbal Tea" &&
        this.drugs[i].name != "Fervex"
      ) {
        if (this.drugs[i].benefit > 0) {
          if (this.drugs[i].name != "Magic Pill") {
            this.drugs[i].benefit = this.drugs[i].benefit - 1;
          }
        }
      } else {
        if (this.drugs[i].benefit < 50) {
          this.drugs[i].benefit = this.drugs[i].benefit + 1;
          if (this.drugs[i].name == "Fervex") {
            if (this.drugs[i].expiresIn < 11) {
              if (this.drugs[i].benefit < 50) {
                this.drugs[i].benefit = this.drugs[i].benefit + 1;
              }
            }
            if (this.drugs[i].expiresIn < 6) {
              if (this.drugs[i].benefit < 50) {
                this.drugs[i].benefit = this.drugs[i].benefit + 1;
              }
            }
          }
        }
      }
      if (this.drugs[i].name != "Magic Pill") {
        this.drugs[i].expiresIn = this.drugs[i].expiresIn - 1;
      }
      if (this.drugs[i].expiresIn < 0) {
        if (this.drugs[i].name != "Herbal Tea") {
          if (this.drugs[i].name != "Fervex") {
            if (this.drugs[i].benefit > 0) {
              if (this.drugs[i].name != "Magic Pill") {
                this.drugs[i].benefit = this.drugs[i].benefit - 1;
              }
            }
          } else {
            this.drugs[i].benefit =
              this.drugs[i].benefit - this.drugs[i].benefit;
          }
        } else {
          if (this.drugs[i].benefit < 50) {
            this.drugs[i].benefit = this.drugs[i].benefit + 1;
          }
        }
      }
    }

    return this.drugs;
  }
}
