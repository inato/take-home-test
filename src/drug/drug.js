import config from "./drugConfig";
export class Drug {
  constructor(name, expiresIn, benefit) {
    this.name = name;
    this.expiresIn = expiresIn;
    this.benefit = this.getBenefitInRange(benefit);
    this.minBenefit = 0;
    this.maxBenefit = 50;

    let conf = config.find(drug => drug.name === name);
    if (!conf) {
      conf = config.find(drug => drug.name === "regular");
    }
    this.config = conf;
  }

  getBenefitInRange(value) {
    return Math.max(0, Math.min(value, 50));
  }

  update() {
    this.updateExpiresIn();
    this.updateBenefit();
  }

  updateExpiresIn() {
    const { expiresInDelta } = this.config;

    const { delta, value } = this.getDeltaOrValueFromRulesArray(expiresInDelta);
    this.expiresIn = typeof value === "number" ? value : this.expiresIn + delta;
  }

  updateBenefit() {
    const { benefitDelta } = this.config;

    const { delta, value } = this.getDeltaOrValueFromRulesArray(benefitDelta);
    const calculatedBenefit =
      typeof value === "number" ? value : this.benefit + delta;
    this.benefit = this.getBenefitInRange(calculatedBenefit);
  }

  getDeltaOrValueFromRulesArray(rules) {
    let d, v;
    for (let { gte, lt, delta, value } of rules) {
      if (this.expiresIn >= gte && this.expiresIn < lt) {
        if (typeof delta === "number") {
          d = delta;
        } else if (typeof value === "number") {
          v = value;
        }
      }
    }
    return { delta: d, value: v };
  }
}
