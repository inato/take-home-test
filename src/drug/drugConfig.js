export default [
  {
    name: "regular",
    expiresInDelta: [
      { gte: Number.NEGATIVE_INFINITY, lt: Number.POSITIVE_INFINITY, delta: -1 }
    ],
    benefitDelta: [
      { gte: 0, lt: Number.POSITIVE_INFINITY, delta: -1 },
      { gte: Number.NEGATIVE_INFINITY, lt: 0, delta: -2 }
    ]
  },
  {
    name: "Herbal Tea",
    expiresInDelta: [
      { gte: Number.NEGATIVE_INFINITY, lt: Number.POSITIVE_INFINITY, delta: -1 }
    ],
    benefitDelta: [
      { gte: 0, lt: Number.POSITIVE_INFINITY, delta: 1 },
      { gte: Number.NEGATIVE_INFINITY, lt: 0, delta: 2 }
    ]
  },
  {
    name: "Magic Pill",
    expiresInDelta: [
      { gte: Number.NEGATIVE_INFINITY, lt: Number.POSITIVE_INFINITY, delta: 0 }
    ],
    benefitDelta: [
      { gte: Number.NEGATIVE_INFINITY, lt: Number.POSITIVE_INFINITY, delta: 0 }
    ]
  },
  {
    name: "Fervex",
    expiresInDelta: [
      { gte: Number.NEGATIVE_INFINITY, lt: Number.POSITIVE_INFINITY, delta: -1 }
    ],
    benefitDelta: [
      { gte: 11, lt: Number.POSITIVE_INFINITY, delta: 1 },
      { gte: 6, lt: 11, delta: 2 },
      { gte: 0, lt: 6, delta: 3 },
      { gte: Number.NEGATIVE_INFINITY, lt: 0, value: 0 }
    ]
  },
  {
    name: "Dafalgan",
    expiresInDelta: [
      { gte: Number.NEGATIVE_INFINITY, lt: Number.POSITIVE_INFINITY, delta: -1 }
    ],
    benefitDelta: [
      { gte: 0, lt: Number.POSITIVE_INFINITY, delta: -2 },
      { gte: Number.NEGATIVE_INFINITY, lt: 0, delta: -4 }
    ]
  }
];
