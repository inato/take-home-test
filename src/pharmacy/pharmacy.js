export class Pharmacy {
  constructor(drugs = []) {
    this.drugs = drugs;
  }
  updateBenefitValue() {
    for (let drug of this.drugs) {
      drug.update();
    }
  }

  getFormattedDrugs() {
    return this.drugs.map(({ name, expiresIn, benefit }) => ({
      name,
      expiresIn,
      benefit
    }));
  }
}
