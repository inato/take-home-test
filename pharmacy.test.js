import { Pharmacy } from "./pharmacy";
import { Drug } from "./drug";

describe("Pharmacy", () => {
  it("should decrease the benefit and expiresIn", () => {
    expect(
      new Pharmacy([new Drug("test", 2, 3)]).updateBenefitValue()
    ).toEqual([new Drug("test", 1, 2)]);
  });
});
