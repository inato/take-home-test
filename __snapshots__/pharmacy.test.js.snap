// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pharmacy Basics should decrease the benefit and expiresIn 1`] = `
Array [
  Drug {
    "benefit": 2,
    "expiresIn": 1,
    "name": "test",
  },
]
`;
